feat(docs): New hybrid mobile agents (.NET MAUI, Xamarin) (#13276)
newrelic/docs-website: src/content/docs/mobile-monitoring/new-relic-mobile-maui-dotnet/monitor-your-net-maui-application.mdx

---
title: Monitor your .NET MAUI app
tags:
    - .NET MAUI
    - Mobile monitoring
    - Hybrid framework
    - Get started
metaDescription: "New Relic's .NET MAUI agent for Android and iOS: Features, compatibility, requirements, installation, and upgrade procedures."
---

import mobileFlutterSummary from 'images/mobile_screenshot-full_flutter-summary.webp'

Monitor your .NET MAUI application with our New Relic .NET MAUI agent. Capture C# errors, track specific network requests, use distributed tracing to drill down into handled exceptions, and create custom tracking events and metrics to fully capture your end users' experience.

<img
  title="Flutter summary view in the UI"
  alt="Summary view of a flutter app in New Relic"
  src={mobileFlutterSummary}
  style={{ maxWidth: '70%' }}
/>

<figcaption>
  **[one.newrelic.com](https://one.newrelic.com/all-capabilities) > Mobile > (select an app) > Summary**: View .NET MAUI data, track HTTP requests and errors, and monitor how your app is performing over time.
</figcaption>

Ready to get started? Follow the steps below! 

## Install the .NET MAUI SDK
<Steps>

<Step>
### Review the requirements

> Before you install the .NET MAUI agent, make sure your .NET MAUI app meets these version requirements:

Author: Is there a .NET MAUI version requirement?

For Android-native apps:
    * Android 7 (API 24) or higher

For iOS-native apps:
    * iOS 11 or higher
    * Latest release of Xcode
</Step>

<Step>
### Add the Flutter agent to your project

> First, you'll need to add the agent package `NewRelic.MAUI.Plugin` to your MAUI project. In your `DIRECTORY`, add the following to `FILE_NAME`: 

Author: Where should they add `NewRelic.MAUI.Plugin` to specifically in their MAUI project?

---
title: Monitor your .NET MAUI mobile application
tags:
    - .NET MAUI
    - Mobile monitoring
    - Hybrid framework
    - Get started

> metaDescription: "New Relic's .NET MAUI agent for Android and iOS: Features, compatibility, requirements, installation, and upgrade procedures."

Reviewer: we use the word 'update' and not 'upgrade', for general description of updating agents and integrations (think it's in style guide)

> metaDescription: "New Relic's .NET MAUI agent for Android and iOS: Features, compatibility, requirements, installation, and upgrade procedures."

Author: ```suggestion
metaDescription: "New Relic's .NET MAUI agent for Android and iOS: Features, compatibility, requirements, and installation."
```

---

import mobileFlutterSummary from 'images/mobile_screenshot-full_flutter-summary.webp'

Our New Relic .NET MAUI agent monitors your .NET MAUI mobile app and provides deep insights into your app's performance, errors, and user experience. With the agent, you can:

* Capture C# errors to quickly identify and fix problems.
* Track network requests to see how your app is interacting with the backend.
* Use distributed tracing to drill down into handled exceptions and identify the root cause of problems.
* Create custom tracking events and metrics to fully understand how your users are interacting with your app.

<img

>   title="Flutter summary view in the UI"
>   alt="Summary view of a flutter app in New Relic"
>   src={mobileFlutterSummary}
>   style={{ maxWidth: '70%' }}

Author: ```suggestion
  title="Mobile summary view in the UI"
  alt="Summary view of a .NET MAUI mobile app in New Relic"
  src={mobileFlutterSummary}
  style={{ maxWidth: '70%' }}
```

/>

<figcaption>
  **[one.newrelic.com](https://one.newrelic.com/all-capabilities) > Mobile > (select your .NET MAUI app) > Summary**: View .NET MAUI data, track HTTP requests and errors, and monitor how your app is performing over time.
</figcaption>

> ## Install the .NET MAUI SDK

Author: ```suggestion
## Install the .NET MAUI SDK [#install]
```

<Steps>

<Step>

> ### Review the requirements

Reviewer: we still want to do the [#requirements] anchor link for these even when in steps, right?

> ### Review the requirements

Author: ```suggestion
### Review the requirements [#requirements]
```

Before you install our .NET MAUI agent, make sure your app meets these version requirements:

> * .NET version 7.0 and higher

Reviewer: ```suggestion
* .NET version 7.0 or higher
```

For Android-native apps:
    * Android 7 (API 24) or higher

For iOS-native apps:
    * iOS 11 or higher
    * Latest release of Xcode
</Step>

<Step>
### Add the .NET MAUI agent to your project

First, you'll need to add our agent, a NuGet package, to your MAUI project:

1. Open your .NET MAUI solution, select the project you want to add the agent to, and open its context menu.
2. Click **Add > Add NuGet packages**, then select `NewRelic.MAUI.Plugin`.

</Step>

<Step>
### Copy your application token from the UI [#app-token]

> The application token is used for New Relic to authenticate your Flutter app's data. 

Reviewer 2: ```suggestion
The application token is used for New Relic to authenticate your .NET MAUI app's data. 
```

> To view and copy your application token in the New Relic UI:

Reviewer: This might be a case where we don't have to document what's already described elsewhere. We could say something like: 

Get your application token from the New Relic mobile UI, on the **Installation** or **Application** page. For more detail, see Application token...

1. In your project, open `App.xaml.cs` and add the following code:

```C#
using NewRelic.MAUI.Plugin;
...
    public App ()
    {
      InitializeComponent();

      MainPage = new AppShell();
      
      CrossNewRelic.Current.HandleUncaughtException();
      CrossNewRelic.Current.TrackShellNavigatedEvents();

      // Set optional agent configuration
      // Options are: crashReportingEnabled, loggingEnabled, logLevel, collectorAddress, crashCollectorAddress
      // AgentStartConfiguration agentConfig = new AgentStartConfiguration(true, true, LogLevel.INFO, "mobile-collector.newrelic.com", "mobile-crash.newrelic.com");

      if (DeviceInfo.Current.Platform == DevicePlatform.Android)
      {
        CrossNewRelic.Current.Start("<APP-TOKEN-HERE>");
        // Start with optional agent configuration 
        // CrossNewRelic.Current.Start("<APP-TOKEN-HERE>", agentConfig);
      } else if (DeviceInfo.Current.Platform == DevicePlatform.iOS)
      {
        CrossNewRelic.Current.Start("<APP-TOKEN-HERE>");
        // Start with optional agent configuration 
        // CrossNewRelic.Current.Start("<APP-TOKEN-HERE>", agentConfig);
      }
    }

```

> 2. In the config file above, replace `<APP-TOKEN-HERE>` with your custom application token(s) (copied from [Step 3]. If you're monitoring both Android-native and iOS-native apps, you'll need to add two different application tokens.

Reviewer 2: I feel calling it `android-native` and `ios-native` on both the .NET MAUI and Xamarin pages is confusing. Maybe a better description would be `Xamarin.Android` & `Xamarin.iOS` for Xamarin and `.NET Android` & `.NET iOS`

> 2. In the config file above, replace `<APP-TOKEN-HERE>` with your custom application token(s) (copied from [Step 3]. If you're monitoring both Android-native and iOS-native apps, you'll need to add two different application tokens.

Author: ```suggestion
2. In the config file above, replace `<APP-TOKEN-HERE>` with your custom application token(s) (copied from [Step 3]. If you're monitoring both `.NET Android` and `.NET iOS` apps, you'll need to add two different application tokens.
```

<Step>

> ### (Android only) Add permissions

Reviewer: ```suggestion

### (Android only) Add permissions
```

* [Crash event trails](/docs/mobile-monitoring/mobile-monitoring-ui/crashes/mobile-crash-event-trail)
* [HTTP requests](/docs/mobile-monitoring/mobile-monitoring-ui/network-pages/http-requests-page)
* [Set up alerts](/docs/mobile-monitoring/new-relic-mobile/get-started/mobile-monitoring-alert-information)

## Troubleshooting

Having trouble finding HTTP data in the UI?

### Problem

After installing the agent and waiting at least 5 minutes, you see **No HTTP data appears:** on the **HTTP errors** and **HTTP requests** UI pages. 

### Solution

To instrument HTTP requests, [make sure you used `HttpMessageHandler` in `HttpClient`](#http-message-handler).

## (Optional) Customize the agent instrumentation

Use the iOS or Android SDK APIs to create custom events and metrics. Here are some examples:

<CollapserGroup>

<Collapser
    id="navigation-events"
    title="Track navigation events"
>
    The .NET MAUI agent makes it easy to track navigation events in a [.NET MAUI shell](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/shell/navigation). To do this, simply call the following code:

```C#
    CrossNewRelic.Current.TrackShellNavigatedEvents();
```

> We recommend you call this method along when starting the agent. These events will only be recorded after navigation is complete. You can find this data through the data explorer in `MobileBreadcrumb` under the name `ShellNavigated`, or by a query such as:

Reviewer: the 'along' seems confusing/unnecessary

``` C#
    CrossNewRelic.Current.NoticeNetworkFailure(
      "https://fakewebsite.com",
      "GET",
      DateTimeOffset.Now.ToUnixTimeMilliseconds(),
      DateTimeOffset.Now.ToUnixTimeMilliseconds() + 100,
      NetworkFailure.Unknown
    );
```
</Collapser>

<Collapser
    id="breadcrumbs"
    title="Record breadcrumbs"
>
    **[RecordBreadcrumb](/docs/mobile-monitoring/new-relic-mobile-android/android-sdk-api/recordbreadcrumb)(string name, Dictionary{'<string, object>'} attributes): bool;**

This call creates and records a `MobileBreadcrumb` event, which can be queried with NRQL and viewed in a crash event trail.

``` C#
    CrossNewRelic.Current.RecordBreadcrumb("MAUIExampleBreadcrumb", new Dictionary{'<string, object>'}()
        {
            {"BreadNumValue", 12.3 },
            {"BreadStrValue", "MAUIBread" },
            {"BreadBoolValue", true }
        }
    );
```
</Collapser>

<Collapser
    id="custom-events"
    title="Record custom events"
>
    **[RecordCustomEvent](/docs/mobile-monitoring/new-relic-mobile-android/android-sdk-api/recordcustomevent-android-sdk-api)(string eventType, string eventName, Dictionary{'<string, object>'} attributes): bool;**

> Creates and records a custom event for use in New Relic Insights.

Author: ```suggestion
Creates and records a custom event for custom data points and dashboards.
```

To view and copy your application token in the New Relic UI:

1. Go to **[one.newrelic.com](https://one.newrelic.com/all-capabilities)**, click **Add data**, then click **Mobile**.
2. Select your .NET MAUI app. 
3. Go to **Settings > Application** and copy the displayed **Application token**.

You'll add this application token in the next step.
</Step>

<Step>
### Add our agent configuration file to your .NET MAUI project [#add-configuration]

1. In your project, open `App.xaml.cs` and add the following code:

```C#
using NewRelic.MAUI.Plugin;
...
    public App ()
    {
      InitializeComponent();

      MainPage = new AppShell();
      
      CrossNewRelic.Current.HandleUncaughtException();
      CrossNewRelic.Current.TrackShellNavigatedEvents();

      // Set optional agent configuration
      // Options are: crashReportingEnabled, loggingEnabled, logLevel, collectorAddress, crashCollectorAddress
      // AgentStartConfiguration agentConfig = new AgentStartConfiguration(true, true, LogLevel.INFO, "mobile-collector.newrelic.com", "mobile-crash.newrelic.com");

      if (DeviceInfo.Current.Platform == DevicePlatform.Android)
      {

>         CrossNewRelic.Current.Start("<APP-TOKEN-HERE>");

Author: ```suggestion
        CrossNewRelic.Current.Start("<YOUR_APP_TOKEN>");
```

* [HTTP requests](/docs/mobile-monitoring/mobile-monitoring-ui/network-pages/http-requests-page)
* [Set up alerts](/docs/mobile-monitoring/new-relic-mobile/get-started/mobile-monitoring-alert-information)

## Troubleshooting [#troubleshooting]

Having trouble finding HTTP data in the UI?

### Problem

After installing the agent and waiting at least 5 minutes, you see **No HTTP data appears:** on the **HTTP errors** and **HTTP requests** UI pages. 

### Solution

To instrument HTTP requests, [make sure you used `HttpMessageHandler` in `HttpClient`](#http-message-handler).

## (Optional) Customize the agent instrumentation [#customize-instrumentation]

Use the iOS or Android SDK APIs to create custom events and metrics. Here are some examples:

<CollapserGroup>

<Collapser
    id="navigation-events"
    title="Track navigation events"
>
    The .NET MAUI agent makes it easy to track navigation events in a [.NET MAUI shell](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/shell/navigation). To do this, simply call the following code:

```C#
    CrossNewRelic.Current.TrackShellNavigatedEvents();
```

> We recommend you call this method along when starting the agent. These events will only be recorded after navigation is complete. You can find this data through the data explorer in `MobileBreadcrumb` under the name `ShellNavigated`, or by a query such as:

Author: ```suggestion
We recommend you call this method immediately after the agent is started. These events will only be recorded after navigation is complete. You can find this data through the data explorer in `MobileBreadcrumb` under the name `ShellNavigated`, or by a query such as:
```

import mobileFlutterSummary from 'images/mobile_screenshot-full_flutter-summary.webp'

Our New Relic .NET MAUI agent monitors your .NET MAUI mobile app and provides deep insights into your app's performance, errors, and user experience. With the agent, you can:

* Capture C# errors to quickly identify and fix problems.
* Track network requests to see how your app is interacting with the backend.
* Use distributed tracing to drill down into handled exceptions and identify the root cause of problems.
* Create custom tracking events and metrics to fully understand how your users are interacting with your app.

<img
  title="Mobile summary view in the UI"
  alt="Summary view of a .NET MAUI mobile app in New Relic"
  src={mobileFlutterSummary}
  style={{ maxWidth: '70%' }}
/>

<figcaption>
  **[one.newrelic.com](https://one.newrelic.com/all-capabilities) > Mobile > (select your .NET MAUI app) > Summary**: View .NET MAUI data, track HTTP requests and errors, and monitor how your app is performing over time.
</figcaption>

## Install the .NET MAUI SDK
<Steps>

<Step>
### Review the requirements

Before you install our .NET MAUI agent, make sure your app meets these version requirements:

> 
> For Android-native apps:
>     * Android 7 (API 24) or higher
> 
> For iOS-native apps:
>     * iOS 11 or higher
>     * Latest release of Xcode

Author: ```suggestion
* .NET version 7.0 and higher
* For Android-native apps: Android 7 (API 24) or higher
* For iOS-native apps:
    * iOS 11 or higher
    * Latest release of Xcode
```

* [Handled exceptions](/docs/mobile-monitoring/mobile-monitoring-ui/crashes/handled-exceptions-analyze-trends-prevent-crashes)
* [Crash event trails](/docs/mobile-monitoring/mobile-monitoring-ui/crashes/mobile-crash-event-trail)
* [HTTP requests](/docs/mobile-monitoring/mobile-monitoring-ui/network-pages/http-requests-page)
* [Set up alerts](/docs/mobile-monitoring/new-relic-mobile/get-started/mobile-monitoring-alert-information)

## Troubleshooting

Having trouble finding HTTP data in the UI?

### Problem

After installing the agent and waiting at least 5 minutes, you see **No HTTP data appears:** on the **HTTP errors** and **HTTP requests** UI pages. 

### Solution

To instrument HTTP requests, [make sure you used `HttpMessageHandler` in `HttpClient`](#http-message-handler).

## Query log data 

New Relic stores your app logs as custom events, which you can query and build dashboards for using the following NRQL query:
```
 SELECT * FROM Mobile Dart Console Events SINCE 30 MINUTES AGO
```

For more information on NRQL queries, see [Introduction to NRQL](/docs/query-your-data/nrql-new-relic-query-language/get-started/introduction-nrql-new-relics-query-language/#where).

> 

Reviewer 3: rightnow there are no logs captured by xamarina and Maui but we will implement it for future

``` C#
    CrossNewRelic.Current.NoticeNetworkFailure(
      "https://fakewebsite.com",
      "GET",
      DateTimeOffset.Now.ToUnixTimeMilliseconds(),
      DateTimeOffset.Now.ToUnixTimeMilliseconds() + 100,
      NetworkFailure.Unknown
    );
```
</Collapser>

<Collapser
    id="breadcrumbs"
    title="Record breadcrumbs"
>
    **[RecordBreadcrumb](/docs/mobile-monitoring/new-relic-mobile-android/android-sdk-api/recordbreadcrumb)(string name, Dictionary{'<string, object>'} attributes): bool;**

This call creates and records a `MobileBreadcrumb` event, which can be queried with NRQL and viewed in a crash event trail.

``` C#
    CrossNewRelic.Current.RecordBreadcrumb("MAUIExampleBreadcrumb", new Dictionary{'<string, object>'}()
        {
            {"BreadNumValue", 12.3 },
            {"BreadStrValue", "MAUIBread" },
            {"BreadBoolValue", true }
        }
    );
```
</Collapser>

<Collapser
    id="custom-events"
    title="Record custom events"
>
    **[RecordCustomEvent](/docs/mobile-monitoring/new-relic-mobile-android/android-sdk-api/recordcustomevent-android-sdk-api)(string eventType, string eventName, Dictionary{'<string, object>'} attributes): bool;**

> Creates and records a custom event for use in New Relic Insights.

Author: @ndesai-newrelic I tried searching for New Relic Insights but couldn't find a good doc to link to.
